File: leetcode-practice-java/src/main/java/io/hank/leetcode/practices/_605_Can_Place_Flowers.java
package io.hank.leetcode.practices;

import io.hank.leetcode.annotations.*;

/**
 * <pre>
 * You have a long flowerbed in which some of the plots are planted, and some are not. However, flowers cannot be planted in adjacent plots.
 *
 * Given an integer array flowerbed containing 0's and 1's, where 0 means empty and 1 means not empty, and an integer n, return true if n new flowers can be planted in the flowerbed without violating the no-adjacent-flowers rule and false otherwise.
 *
 *
 *
 * Example 1:
 *
 * Input: flowerbed = [1,0,0,0,1], n = 1
 * Output: true
 * Example 2:
 *
 * Input: flowerbed = [1,0,0,0,1], n = 2
 * Output: false
 *
 *
 * Constraints:
 *
 * 1 <= flowerbed.length <= 2 * 10^4
 * flowerbed[i] is 0 or 1.
 * There are no two adjacent flowers in flowerbed.
 * 0 <= n <= flowerbed.length
 * </pre>
 */
public class _605_Can_Place_Flowers extends LeetcodeProblemSolution {

    @Topic(TopicType.GREEDY)
    @TimeComplexity(ComplexityType.O_N)
    @SpaceComplexity(ComplexityType.O_1)
    boolean canPlaceFlowers(int[] flowerbed, int n) {
        // check if left is 0 and right is 0 and self is 0, then mark self as 1
        // append both 0 in left of array and in right of array
        // i is self pointer
        for (int i = 0; i < flowerbed.length; i++) {
            if (flowerbed[i] == 1) {
                continue;
            }
            boolean left = i == 0 || flowerbed[i - 1] == 0;
            boolean right = i == flowerbed.length - 1 || flowerbed[i + 1] == 0;
            if (left && right) {
                flowerbed[i] = 1;
                n--;
            }
        }
        return n <= 0;
    }
}

File: leetcode-practice-typescript/src/greedy/605-can-place-flowers.ts
function canPlaceFlowers(flowerbed: number[], n: number): boolean {
    if (n === 0) {
        return true;
    }
    for (let i = 0; i < flowerbed.length; i++) {
        const afterZero: boolean = i === 0 || flowerbed[i - 1] === 0;
        const beforeZero: boolean =
            i === flowerbed.length - 1 || flowerbed[i + 1] === 0;
        if (flowerbed[i] === 0 && afterZero && beforeZero) {
            n--;
            flowerbed[i] = 1;
            if (n === 0) {
                return true;
            }
        }
    }
    return false;
}

canPlaceFlowers([1, 0, 0, 0, 1], 1);
